### PR TITLE
DOC: Add Legacy directive

### PR DESCRIPTION
**Pull request by melissawm**
Implements a legacy directive to be used in the documentation. 

This has been discussed in the community meetings. See also [this discussion on the mailing list](https://mail.python.org/archives/list/scipy-dev@python.org/thread/Z33JO27EX6BX7LVBI6CW6MX64VQZIJ66/#6GLVRHX5XBTZNL62LFN4YPGACY5BDPLA).

#### What does this implement/fix?
Creates a custom Sphinx `Admonition` to be used as a note in functions, modules or pages that are considered legacy. 

Example usage:

```
.. legacy::

   New code should use :mod:`scipy.fft`.
```

The result is

Screenshot_20221214_085149

Both styling and implementation could be improved, feedback is welcome!

### CODE REVIEW

**Comment by tupui** [comment]
Thanks @melissawm. This looks very similar to the admonitions in the sphinx theme, see https://pydata-sphinx-theme.readthedocs.io/en/stable/examples/kitchen-sink/admonitions.html

Is there a difference with what you want to achieve?

**Comment by ev-br** [comment]
Cannot comment on the implementation, but for the usage I would love to use it in several places!

**Comment by melissawm** [comment]
Ok - I found a better solution!

This will now do what I wanted which is:

```
.. legacy::
```

with no arguments will generate the default text; if we do add any content, it will be appended to the default text. 

I didn't change the styling yet, and I'm not the best at design so I'll take any suggestions for colors or styles. I've also added some documentation on the usage but I'm not sure if this is the right place to add.

**Comment by melissawm** [comment]
Looks like I also need to update the refguide_check.

**Comment by melissawm** [comment]
I think it worked - @ev-br would you mind taking a look at it just to make sure this is reasonable? Thanks!

**Comment by ev-br** [comment]
It does seem to work, great!

https://output.circle-artifacts.com/output/job/5bb2caee-59bf-46ce-97f8-51e17eb9413d/artifacts/0/html/reference/fftpack.html

Two minor comments:

- color scheme. Should it be reddish like `.. warning::` ? The current box is too easy to miss.
- Is it correct that whatever is written explicitly gets appended to "This submodule is deprecated"? As a use case, what's the correct way of adding this directive to a single class/function in a submodule? (the case in point is `interp1d` or `Rbf`: both are legacy but `scipy.interpolate` isn't)

**Comment by ev-br** [comment]
I'm happy to leave the decision on the color scheme and the exact wording to others; would only note we probably want one admonition, not two. 

The default _option_ looks good to me if it's not difficult. Having no default is also an alternative IMO --- as long as it allows to annotate an individual function, it works :-).

**Comment by ev-br** [comment]
However much I hate to be a pain in the neck, the wording in https://output.circle-artifacts.com/output/job/4d0c259c-84cb-43e5-8d75-a23b9976d8c3/artifacts/0/html/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d

does read a bit strange:

> This submodule is now considered legacy and will no longer receive updates. function

**Comment by mdhaber** [comment]
This looks great!

Re: https://github.com/scipy/scipy/pull/17597#issuecomment-1371049444, I agree that the new color suggestion will stand out better than the current one. Regarding deprecation, the idea is that that you'd use the `.. legacy::` directive with `object` as an argument so that you don't also need a separate `.. deprecated::` directive? 

Re: https://github.com/scipy/scipy/pull/17597#issuecomment-1374730203, I think it's just "function" at the end that

**Comment by melissawm** [comment]
> However much I hate to be a pain in the neck, the wording in [output.circle-artifacts.com/output/job/4d0c259c-84cb-43e5-8d75-a23b9976d8c3/artifacts/0/html/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d](https://output.circle-artifacts.com/output/job/4d0c259c-84cb-43e5-8d75-a23b9976d8c3/artifacts/0/html/reference/generated/scipy.interpolate.interp1d.html#scipy.interpolate.interp1d)
> 
> does read a bit strange:
> 
> > This submodule is now considered legacy and will no longer receive updates. function

oops. Something went wrong here. Let me fix it

**Comment by melissawm** [comment]
> Re: [#17597 (comment)](https://github.com/scipy/scipy/pull/17597#issuecomment-1371049444), I agree that the new color suggestion will stand out better than the current one. Regarding deprecation, the idea is that that you'd use the `.. legacy::` directive with `object` as an argument so that you don't also need a separate `.. deprecated::` directive?


I had not thought of the deprecated directive - as I understand this is a different idea, not necessarily meant for deprecations. But maybe @rgommers had something different in mind? I'm happy to make these work together though.

**Comment by rgommers** [comment]
> I had not thought of the deprecated directive - as I understand this is a different idea, not necessarily meant for deprecations. But maybe @rgommers had something different in mind? I'm happy to make these work together though.

They're separate directives, no cross-talk there. Legacy means "there's a better alternative by now, but we're _not_ deprecating it - it's here for backwards compatibility". There were more details on the mailing list.

**Comment by mdhaber** [comment]
I don't think we need to combine them. It seems that "deprecated" is a step down from "legacy" - not only will the feature not be maintained, but removal is planned. If something that is marked as legacy is deprecated, I think we can *replace* the `..legacy` directive with the `..deprecated` directive.

**Comment by ev-br** [comment]
What's the status here, is there something to do still? Otherwise, I think it'd be great to have, I've a few uses :-)

**Comment by melissawm** [comment]
I think this is done 🎉 happy to get feedback on potential easier/simpler ways to solve this, but I couldn't find any. Using a basic custom admonition [like this one](https://github.com/executablebooks/meta/discussions/655) I couldn't add the extra default text; using an "attention"-class admonition there was no way to add a custom title. Creating a custom admonition and a custom new class for styling (`admonition-legacy`) solves both issues, so this is how it looks like now:

Screenshot_20230319_171013

**Comment by melissawm** [comment]
(Looks like the lint is picking up some stuff that doesn't belong to this PR. I'm happy to fix those on a separate commit before this gets merged if it helps)

File: scipy/interpolate/_interpolate.py
Comment (tupui) [inline scipy/interpolate/_interpolate.py:388]: Nitpick, but this is a class. From the code above, I am not sure if it works with a class. It should right?

File: doc/source/conf.py
Comment (tupui) [inline doc/source/conf.py:421]: Can be done in a follow up, but I would move this in its own file to not mix it with the config itself. Otherwise we are missing 2 spaces before it.
```suggestion


class LegacyDirective(Directive):
```

File: doc/source/dev/contributor/rendering_documentation.rst
Comment (tupui) [inline doc/source/dev/contributor/rendering_documentation.rst:304]: I would actually consider to force us to write something. That would help users have clear update paths.

If we don't want to force that, we could at least mention that it's recommended to do so.

File: doc/source/conf.py
Comment (tupui) [inline doc/source/conf.py:439]: I don't think `now` brings useful information.

```suggestion
        text = (f"This {obj} is considered legacy and will no longer "
```

File: doc/source/conf.py
Comment (tupui) [inline doc/source/conf.py:440]: Should we clarify that, not receiving updates could mean a removal at some point?

In a follow up we should have a section about legacy and deprecation clarifying our policy. (Could be alongside the existing section we have in the core dev part of the contributing guide.)

**Review by tupui — APPROVED** [review]
Thank you @melissawm! I have some minor non blocking comments.

File: doc/source/conf.py
Comment (melissawm) [reply doc/source/conf.py:421]: Adding the spaces for now but happy to move it to another file in a follow up, together with the extra docs you mentioned below!

File: scipy/interpolate/_interpolate.py
Comment (melissawm) [reply scipy/interpolate/_interpolate.py:388]: It should! Updated it, thanks

File: doc/source/dev/contributor/rendering_documentation.rst
Comment (melissawm) [reply doc/source/dev/contributor/rendering_documentation.rst:304]: Let me know if this wording is ok.

**Comment by melissawm** [comment]
Rebased to fix conflict.

**Comment by tupui** [comment]
> Rebased to fix conflict.

No worries 😃 next time please prefer a merge or be just to check with other maintainers before doing it. The issue is that it kind of breaks the "review history". e.g. if you go to the review tab and look at the conversations, the links don't work anymore. Yeah GH is not really helpful here 😅 

Screenshot 2023-03-22 at 23 46 15

**Review by tupui — APPROVED** [review]
Thank you for the update @melissawm 🙌 LGTM and we discussed it during todays meeting and concluded it was a great addition 😃

**Pull request merged**
